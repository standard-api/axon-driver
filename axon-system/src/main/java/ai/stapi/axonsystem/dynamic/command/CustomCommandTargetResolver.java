package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.AbstractCommand;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.modelling.command.AnnotationCommandTargetResolver;
import org.axonframework.modelling.command.CommandTargetResolver;
import org.axonframework.modelling.command.VersionedAggregateIdentifier;
import org.jetbrains.annotations.NotNull;

public class CustomCommandTargetResolver implements CommandTargetResolver {
  
  private final AnnotationCommandTargetResolver annotationCommandTargetResolver;

  public CustomCommandTargetResolver() {
    this.annotationCommandTargetResolver = AnnotationCommandTargetResolver.builder().build();
  }

  @Override
  public VersionedAggregateIdentifier resolveTarget(@NotNull CommandMessage<?> command) {
    var payload = command.getPayload();
    if (payload instanceof AbstractCommand<?> abstractCommand) {
      return new VersionedAggregateIdentifier(
          abstractCommand.getTargetIdentifier(),
          null
      );
    }
    return this.annotationCommandTargetResolver.resolveTarget(command);
  }
}
