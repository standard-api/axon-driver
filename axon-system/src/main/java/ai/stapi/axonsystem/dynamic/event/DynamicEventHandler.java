package ai.stapi.axonsystem.dynamic.event;

import ai.stapi.axonsystem.dynamic.messagehandler.DynamicMessageHandler;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import org.axonframework.eventhandling.EventMessage;

@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.ANNOTATION_TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD})
@DynamicMessageHandler(
    messageType = EventMessage.class,
    payloadType = DynamicGraphUpdatedEvent.class
)
public @interface DynamicEventHandler {

  /**
   * Specifies the name of message that can be handled by the member method.
   */
  String messageName();

}
