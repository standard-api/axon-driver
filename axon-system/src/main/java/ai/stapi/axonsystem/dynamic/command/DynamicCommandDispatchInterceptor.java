package ai.stapi.axonsystem.dynamic.command;

import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import java.util.List;
import java.util.function.BiFunction;
import org.axonframework.commandhandling.CommandMessage;
import org.axonframework.commandhandling.GenericCommandMessage;
import org.axonframework.messaging.GenericMessage;
import org.axonframework.messaging.MessageDispatchInterceptor;
import org.jetbrains.annotations.NotNull;

public class DynamicCommandDispatchInterceptor  implements MessageDispatchInterceptor<CommandMessage<?>> {

  @NotNull
  @Override
  public BiFunction<Integer, CommandMessage<?>, CommandMessage<?>> handle(
      @NotNull List<? extends CommandMessage<?>> messages
  ) {
    return (integer, commandMessage) -> {
      if (commandMessage.getPayload() instanceof DynamicCommand dynamicCommand) {
        return new GenericCommandMessage<>(
            new GenericMessage<>(dynamicCommand),
            dynamicCommand.getSerializationType()
        );
      }
      return commandMessage;
    };
  }
}
