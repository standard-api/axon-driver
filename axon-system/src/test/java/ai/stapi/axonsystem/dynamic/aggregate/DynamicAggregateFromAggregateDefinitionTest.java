package ai.stapi.axonsystem.dynamic.aggregate;

import ai.stapi.axonsystem.dynamic.aggregate.config.DynamicAggregateFromDefinitionTestConfiguration;
import ai.stapi.axonsystem.dynamic.aggregate.testImplementations.TestAggregateDefinitionProvider;
import ai.stapi.graphsystem.messaging.command.DynamicCommand;
import ai.stapi.graphsystem.messaging.event.DynamicGraphUpdatedEvent;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.test.DomainTestCase;
import java.util.Map;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.context.annotation.Import;

@Import(DynamicAggregateFromDefinitionTestConfiguration.class)
class DynamicAggregateFromAggregateDefinitionTest extends DomainTestCase {

  @Test
  void itCanCreateDynamicAggregate() {
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CREATE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleValueSet"
            )
        )
    );
    var event = this.getLastEventOfType(DynamicGraphUpdatedEvent.class);
    Assertions.assertEquals(
        TestAggregateDefinitionProvider.VALUE_SET_CREATED_EVENT,
        event.getEventName()
    );
    this.thenLastEventOfTypeGraphApproved(DynamicGraphUpdatedEvent.class);

  }

  @Test
  void itCanChangeDynamicAggregate() {
    this.givenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CREATE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleValueSet"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CHANGE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleChangedValueSet"
            )
        )
    );
    this.thenMergedGraphOfAggregateApproved("ExampleValueSetId");
  }

  @Test
  void itCanCreateTwoAggregatesOfSameType() {
    this.givenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CREATE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleValueSet"
            )
        )
    );
    this.givenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId2"),
            TestAggregateDefinitionProvider.CREATE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleValueSet2"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CHANGE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleChangedValueSet"
            )
        )
    );
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId2"),
            TestAggregateDefinitionProvider.CHANGE_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleChangedValueSet2"
            )
        )
    );
    this.thenMergedGraphOfAggregateApproved("ExampleValueSetId");
  }

  @Test
  void itCanCreateDynamicAggregateIfMissing() {
    this.whenCommandIsDispatched(
        new DynamicCommand(
            new UniqueIdentifier("ExampleValueSetId"),
            TestAggregateDefinitionProvider.CREATE_IF_MISSING_VALUE_SET_COMMAND,
            Map.of(
                "name", "ExampleCreatedValueSet"
            )
        )
    );
    this.thenLastEventOfTypeGraphApproved(DynamicGraphUpdatedEvent.class, 1);
  }
}