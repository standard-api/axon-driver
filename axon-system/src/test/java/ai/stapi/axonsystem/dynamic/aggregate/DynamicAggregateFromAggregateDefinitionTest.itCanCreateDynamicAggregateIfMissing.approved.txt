nodes:
	node_type: ValueSet
	node_hash: 9BEB3
	node_edges:
	attributes:
		name -> ExampleChangedValueSet

edges:
